DEVELOPING.md

```diff
@@ -53,3 +53,78 @@ When designing a forest for a new statistical task, we suggest following the for
+
+Two added features to grf's splitting rule beyond basic CART is the addition of sample weights and support for missing values with Missingness Incorporated in Attributes (MIA) splitting.
+
+***Algorithm*** (`RegressionSplittingRule`): find the best split for variable `var` at n samples `samples = i...j`
+
+_Input_: `X`: covariate matrix, `weights`: weight n-vector, `response`: the responses (pseudo-outcomes) n-vector, `min_child_size`: split constraint on number of samples
+
+_Output_: The best split value and the direction to send missing values
+
+```
+total_sum = sum(response * weights)
+total_weight_sum = sum(weights)
+total_samples = count(samples)
+
+// Sort everything according to feature `var`
+sort(X[samples, var], weights, response)
+
+// `unique_cum_sum(x)` computes the cumulative sum for each of the unique elements in X[samples, var]
+// Ex: `unique_cum_sum([1.5, 2.5, 3, 8])` at x-values [0, 0, 0.5, 0.9] is [4, 7, 15]
+// (analogous for count, each of these vectors have length equal to the number of unique samples)
+
+weight_sums_left = unique_cum_sum(weights[samples != "missing"])
+sample_counters_left = unique_cum_count(X[samples != "missing", var])
+weighted_response_sums_left = unique_cum_sum(response[samples != "missing"]
+                                             * weights[samples != "missing"])
+
+n_missing = count(X[samples = "missing", var])
+weight_sum_missing = sum(weights[samples = "missing", var])
+weighted_response_sum_missing = sum(response[samples = "missing"]
+                                    * weights[samples = "missing"])
+
+// a) Calculate the decrease in impurity "sum_Y_left^2/weight_sum_left+sum_Y_right^2/weight_sum_right"
+// first with all missing sent right
+missing_direction = right
+
+weight_sums_right = total_weight_sum - weight_sums_left
+sample_counters_right = total_samples - sample_counters_left
+weighted_response_sums_right = total_sum - weighted_response_sums_left
+
+decrease_na_right = weighted_response_sums_left^2 / weight_sums_left +
+                    weighted_response_sums_right^2 / weight_sums_right
+
+// select the maximum which satisfies the minimum child size constraint, etc:
+max, arg.max = max(decrease_na_left[sample_counters_left >= min_child_size
+                                    AND sample_counters_right >= min_child_size])
+
+// b) The same decrease calculation, but with all missing sent left
+weight_sums_left += weight_sum_missing
+sample_counters_left += n_missing
+weighted_response_sums_left += weighted_response_sum_missing
+
+weight_sums_right = total_weight_sum - weight_sums_left
+sample_counters_right = total_samples - sample_counters_left
+weighted_response_sums_right = total_sum - weighted_response_sums_left
+
+decrease_na_left = weighted_response_sums_left^2 / weight_sums_left +
+                   weighted_response_sums_right^2 / weight_sums_right
+
+if max(decrease_na_left) > max(decrease_na_right):
```

Comment:
This pseudocode was hard for me to understand. The detailed computations obscure the meaning of what is happening. I think something like the following would be easier on readers:

for each unique value u in X
   let  count_left[u] be the number of samples with X <= u 
   let  sum_left[u], weight_sum_left[u] be the sum of weight*response and weight over those elements
   and let *_right[u], *_missing be the same for samples with X > u and X missing.
   
   decrease[u, missing=on-right] = sum_left[u]^2 / weight_sum_left[u] + sum_right[u]^2 / weight_sums_right[u]
   decrease[u, missing=on-left] = ...

return (u, missing) that maximizing decrease among u with count_left[u] and count_right[u] >= min-child-size

It may also help to have an explanation for why decrease has the form it does, with a square in the numerator and none in the denominator.

Reply:
Thanks, it's been simplified. Do you have a suggestion for a decrease explanation?

Reply:
I'll add a decrease explanation here over the weekend, thanks @halflearned for the algebra clarification.

Reply:
A decrease explanation is now added to DEVELOPING, if this is OK I'll go ahead and merge.